tls/s2n_x509_validator.c

```diff
@@ -567,54 +568,62 @@ static S2N_RESULT s2n_x509_validator_set_no_check_time_flag(struct s2n_x509_vali
     return S2N_RESULT_OK;
 }
 
-int s2n_disable_time_validation_ossl_verify_callback(int default_ossl_ret, X509_STORE_CTX *ctx)
+bool s2n_x509_validator_is_ossl_cert_time_validation_error(X509_STORE_CTX *ctx)
 {
     int err = X509_STORE_CTX_get_error(ctx);
     switch (err) {
         case X509_V_ERR_CERT_NOT_YET_VALID:
         case X509_V_ERR_CERT_HAS_EXPIRED:
-            return OSSL_VERIFY_CALLBACK_IGNORE_ERROR;
+            return true;
         default:
-            break;
+            return false;
     }
+}
 
-    /* If CRL validation is enabled, setting the time validation verify callback will override the
-     * CRL verify callback. The CRL verify callback is manually triggered to work around this
-     * issue.
-     *
-     * The CRL verify callback ignores validation errors exclusively for CRL timestamp fields. So,
-     * if CRL validation isn't enabled, the CRL verify callback is a no-op.
-     */
-    return s2n_crl_ossl_verify_callback(default_ossl_ret, ctx);
+bool s2n_x509_validator_is_ossl_crl_time_validation_error(X509_STORE_CTX *ctx)
+{
+    int err = X509_STORE_CTX_get_error(ctx);
+    switch (err) {
+        case X509_V_ERR_CRL_NOT_YET_VALID:
+        case X509_V_ERR_CRL_HAS_EXPIRED:
+        case X509_V_ERR_ERROR_IN_CRL_LAST_UPDATE_FIELD:
+        case X509_V_ERR_ERROR_IN_CRL_NEXT_UPDATE_FIELD:
+            return true;
+        default:
+            return false;
+    }
 }
 
-static S2N_RESULT s2n_x509_validator_disable_time_validation(struct s2n_connection *conn,
-        struct s2n_x509_validator *validator)
+/* Setting an X509_STORE verify callback is generally not recommended with AWS-LC:
+ * https://github.com/aws/aws-lc/blob/aa90e509f2e940916fbe9fdd469a4c90c51824f6/include/openssl/x509.h#L2980-L2990
+ *
+ * Functionality should be implemented using the verify callback only when no alternative configuration for
+ * X509_STORE_CTX exists. Additionally, implementations should aim for minimal API usage to lessen the risk of breaking
+ * libcrypto changes.
+ */
+int s2n_ossl_verify_callback(int default_ossl_ret, X509_STORE_CTX *ctx)
 {
-    RESULT_ENSURE_REF(conn);
-    RESULT_ENSURE_REF(conn->config);
-    RESULT_ENSURE_REF(validator);
-    RESULT_ENSURE_REF(validator->store_ctx);
+    const struct s2n_connection *conn = (struct s2n_connection *) X509_STORE_CTX_get_app_data(ctx);
+    if (conn == NULL || ctx == NULL) {
```

Comment:
It might be good to check ctx before giving it to the libcrypto in `X509_STORE_CTX_get_app_data()`.